File: app/src/androidTest/java/com/zhiyong/tingxie/RecyclerViewTest.java
package com.zhiyong.tingxie;

import androidx.annotation.NonNull;
import androidx.test.core.app.ApplicationProvider;
import androidx.test.espresso.ViewInteraction;
import androidx.test.espresso.contrib.PickerActions;
import androidx.test.espresso.contrib.RecyclerViewActions;
import androidx.test.espresso.matcher.BoundedMatcher;
import androidx.test.filters.LargeTest;
import androidx.test.rule.ActivityTestRule;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;
import android.widget.DatePicker;

import com.zhiyong.tingxie.ui.main.MainActivity;

import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.Matchers;
import org.hamcrest.TypeSafeMatcher;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.Calendar;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.swipeLeft;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.core.internal.deps.guava.base.Preconditions.checkNotNull;
import static androidx.test.espresso.matcher.ViewMatchers.hasDescendant;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withClassName;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static org.hamcrest.CoreMatchers.containsString;
import static org.hamcrest.Matchers.allOf;

@LargeTest
@RunWith(AndroidJUnit4.class)
public class RecyclerViewTest {
    @Rule
    public ActivityTestRule<MainActivity> mActivityTestRule = new ActivityTestRule<>(MainActivity.class);

    private ViewInteraction floatingActionButton = getFAB();

    @Before
    public void before() {
        ApplicationProvider.getApplicationContext().deleteDatabase("pinyin_database");
    }

    @Test
    public void addQuizInFuture() {
        Calendar c = Calendar.getInstance();
        c.add(Calendar.DATE, 1);
        int year = c.get(Calendar.YEAR);
        int monthOfYear = c.get(Calendar.MONTH);
        int dayOfMonth = c.get(Calendar.DATE);
        addQuiz(year, monthOfYear, dayOfMonth);

        onView(withId(R.id.recyclerview_main))
                .check(matches(atPosition(0, hasDescendant(withText(containsString(
                        String.valueOf(dayOfMonth)
                ))))));

        onView(withId(R.id.recyclerview_main)).perform(
                RecyclerViewActions.actionOnItemAtPosition(0, swipeLeft()));
    }

    @Test
    public void quizzesSortedAfterAdd() {
        addQuiz(2019, 3, 1);
        addQuiz(2019, 2, 28);
        onView(withId(R.id.recyclerview_main))
                .check(matches(atPosition(0, hasDescendant(withText(containsString(
                        "01 Mar 2019"))))));
        onView(withId(R.id.recyclerview_main))
                .check(matches(atPosition(1, hasDescendant(withText(containsString(
                        "28 Feb 2019"))))));
        removeQuiz(0);
        removeQuiz(0);
    }

    @Test
    public void quizzesSortedAfterEdit() {
        addQuiz(2019, 3, 1);
        addQuiz(2019, 3, 2);

        onView(withId(R.id.recyclerview_main))
                .check(matches(atPosition(0, hasDescendant(withText(containsString(
                        "02 Mar 2019"))))));
        onView(withId(R.id.recyclerview_main))
                .check(matches(atPosition(1, hasDescendant(withText(containsString(
                        "01 Mar 2019"))))));
        onView(withId(R.id.recyclerview_main)).perform(
                RecyclerViewActions.actionOnItemAtPosition(
                        0, MyViewAction.clickChildViewWithId(R.id.ivEditDate)
                )
        );

        onView(withClassName(Matchers.equalTo(DatePicker.class.getName())))
                .perform(PickerActions.setDate(2019, 4, 2));
        onView(withId(android.R.id.button1)).perform(click());

        onView(withId(R.id.recyclerview_main))
                .check(matches(atPosition(0, hasDescendant(withText(containsString(
                        "02 Apr 2019"))))));
        onView(withId(R.id.recyclerview_main))
                .check(matches(atPosition(1, hasDescendant(withText(containsString(
                        "01 Mar 2019"))))));

        removeQuiz(0);
        removeQuiz(0);
    }

    private void addQuiz(int year, int monthOfYear, int dayOfMonth) {
        floatingActionButton.perform(click());
        onView(withClassName(Matchers.equalTo(DatePicker.class.getName())))
                .perform(PickerActions.setDate(year, monthOfYear, dayOfMonth));
        onView(withId(android.R.id.button1)).perform(click());
    }

    private void removeQuiz(int position) {
        onView(withId(R.id.recyclerview_main)).perform(
                RecyclerViewActions.actionOnItemAtPosition(position, swipeLeft()));
    }

    private ViewInteraction getFAB() {
        ViewInteraction floatingActionButton = onView(
                allOf(withId(R.id.fab),
                        childAtPosition(
                                childAtPosition(
                                        withId(android.R.id.content),
                                        0),
                                2),
                        isDisplayed()));
        return floatingActionButton;
    }

    private static Matcher<View> childAtPosition(
            final Matcher<View> parentMatcher, final int position) {

        return new TypeSafeMatcher<View>() {
            @Override
            public void describeTo(Description description) {
                description.appendText("Child at position " +
                        position + " in parent ");
                parentMatcher.describeTo(description);
            }

            @Override
            public boolean matchesSafely(View view) {
                ViewParent parent = view.getParent();
                return parent instanceof ViewGroup
                        && parentMatcher.matches(parent)
                        && view.equals(((ViewGroup) parent).getChildAt(position));
            }
        };
    }

    public static Matcher<View> atPosition(final int position, @NonNull final Matcher<View> itemMatcher) {
        checkNotNull(itemMatcher);
        return new BoundedMatcher<View, RecyclerView>(RecyclerView.class) {
            @Override
            public void describeTo(Description description) {
                description.appendText("has item at position " + position + ": ");
                itemMatcher.describeTo(description);
            }

            @Override
            protected boolean matchesSafely(final RecyclerView view) {
                RecyclerView.ViewHolder viewHolder = view.findViewHolderForAdapterPosition(position);
                if (viewHolder == null) {
                    // has no item on such position
                    return false;
                }
                return itemMatcher.matches(viewHolder.itemView);
            }
        };
    }

//        ViewInteraction textView = onView(
//                allOf(withId(R.id.word), withText("+ Word 20"),
//                        childAtPosition(
//                                childAtPosition(
//                                        withId(R.id.recyclerview_main),
//                                        13),
//                                0),
//                        isDisplayed()));
//        textView.check(matches(withText("+ Word 20")));
//    MainActivity activity = mActivityTestRule.getActivity();
//    RecyclerView recyclerView = activity.findViewById(R.id.recyclerview_main);
}

File: app/src/androidTest/java/com/zhiyong/tingxie/RecyclerViewTest.kt
package com.zhiyong.tingxie

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.DatePicker
import androidx.recyclerview.widget.RecyclerView
import androidx.test.core.app.ApplicationProvider
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.ViewInteraction
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.contrib.PickerActions
import androidx.test.espresso.contrib.RecyclerViewActions
import androidx.test.espresso.core.internal.deps.guava.base.Preconditions
import androidx.test.espresso.matcher.BoundedMatcher
import androidx.test.espresso.matcher.ViewMatchers.*
import androidx.test.ext.junit.rules.ActivityScenarioRule
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.LargeTest
import com.zhiyong.tingxie.ui.main.MainActivity
import kotlinx.coroutines.runBlocking
import org.hamcrest.*
import org.hamcrest.CoreMatchers.containsString
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import java.text.SimpleDateFormat
import java.util.*


@LargeTest
@RunWith(AndroidJUnit4::class)
class RecyclerViewTest {
  private val floatingActionButton = fAB

  @get:Rule
  val activityRule = ActivityScenarioRule(MainActivity::class.java)

  @Before
  fun before() {
    ApplicationProvider.getApplicationContext<Context>()
      .deleteDatabase("pinyin_database")
//    launch(MainActivity::class.java)
    removeAllQuizzes(activityRule)
  }

  @After
  fun after() {
//    removeAllQuizzes()
  }

  @Test
  fun addQuizInFuture() {
    val c = Calendar.getInstance()
    c.add(Calendar.DATE, 1)
    addQuiz(c)
    checkMatches(c, 0)
  }

  @Test
  fun quizzesSortedAfterAdd() {
    val c = Calendar.getInstance()
    addQuiz(c)
    c.add(Calendar.DATE, 1)
    addQuiz(c)
    checkMatches(c, 0)
    c.add(Calendar.DATE, -1)
    checkMatches(c, 1)
  }

  @Test
  fun quizzesSortedAfterEdit() {
    val c = Calendar.getInstance()
    addQuiz(c)
    c.add(Calendar.DATE, 1)
    addQuiz(c)
    checkMatches(c, 0)
    c.add(Calendar.DATE, -1)
    checkMatches(c, 1)
    onView(withId(R.id.recyclerview_main)).perform(
      RecyclerViewActions.actionOnItemAtPosition<RecyclerView.ViewHolder>(
        0, MyViewAction.clickChildViewWithId(R.id.ivEditDate)
      )
    )

    c.add(Calendar.MONTH, 1)
    // Edit quiz date.
    onView(
      withClassName(
        Matchers.equalTo(
          DatePicker::class.java.name
        )
      )
    ).perform(
      PickerActions.setDate(
        c[Calendar.YEAR],
        c[Calendar.MONTH] + 1,
        c[Calendar.DATE]
      )
    )
    onView(withId(android.R.id.button1)).perform(ViewActions.click())

    checkMatches(c, 0)
    c.add(Calendar.MONTH, -1)
    checkMatches(c, 1)
  }

  private fun addQuiz(year: Int, monthOfYear: Int, dayOfMonth: Int) {
    floatingActionButton.perform(ViewActions.click())
    onView(
      withClassName(
        Matchers.equalTo(
          DatePicker::class.java.name
        )
      )
    ).perform(PickerActions.setDate(year, monthOfYear, dayOfMonth))
    onView(withId(android.R.id.button1)).perform(ViewActions.click())
  }

  private fun addQuiz(c: Calendar) =
    addQuiz(c[Calendar.YEAR], c[Calendar.MONTH] + 1, c[Calendar.DATE])

  private val fAB: ViewInteraction
    get() = onView(
      Matchers.allOf(
        withId(R.id.fab), childAtPosition(
          childAtPosition(withId(android.R.id.content), 0), 2
        ), isDisplayed()
      )
    )

  companion object {
    private fun childAtPosition(
      parentMatcher: Matcher<View>, position: Int
    ): Matcher<View> {
      return object : TypeSafeMatcher<View>() {
        override fun describeTo(description: Description) {
          description.appendText(
            "Child at position " + position + " in parent "
          )
          parentMatcher.describeTo(description)
        }

        public override fun matchesSafely(view: View): Boolean {
          val parent = view.parent
          return (parent is ViewGroup && parentMatcher.matches(parent) && view == parent.getChildAt(
            position
          ))
        }
      }
    }

    fun removeAllQuizzes(activityRule: ActivityScenarioRule<MainActivity>) {
      var count = 0
      runBlocking {
        activityRule.scenario.onActivity {
          val recyclerView = it.findViewById<RecyclerView>(R.id.recyclerview_main)
          count = recyclerView.adapter?.itemCount ?: 0
        }
      }
      try {
        repeat(count) {
          onView(withId(R.id.recyclerview_main)).perform(
            RecyclerViewActions.actionOnItemAtPosition<RecyclerView.ViewHolder>(
              0, ViewActions.swipeLeft()
            )
          )
        }
      } catch (_: Exception) {

      }
      val recyclerView = onView(withId(R.id.recyclerview_main))
//      recyclerView.check(RecyclerViewItemCountAssertion(0))
    }

    private fun atPosition(
      position: Int, itemMatcher: Matcher<View?>
    ): BoundedMatcher<View?, RecyclerView> {
      Preconditions.checkNotNull(itemMatcher)
      return object : BoundedMatcher<View?, RecyclerView>(RecyclerView::class.java) {
        override fun describeTo(description: Description) {
          description.appendText("has item at position $position: ")
          itemMatcher.describeTo(description)
        }

        override fun matchesSafely(view: RecyclerView): Boolean {
          val viewHolder = view.findViewHolderForAdapterPosition(position)
            ?: // has no item on such position
            return false
          return itemMatcher.matches(viewHolder.itemView)
        }
      }
    }

    fun checkMatches(c: Calendar, position: Int) {
      onView(withId(R.id.recyclerview_main)).check(
        matches(
          atPosition(
            position,
            hasDescendant(
              withText(
                containsString(
                  SimpleDateFormat("EEE, dd MMM yyyy").format(c.time)
                )
              )
            )
          )
        )
      )
    }

//    fun checkMatchString(string: String): (Int) -> Unit =
//      { position: Int -> checkMatches(position, string) }
  }
}